packages/grafana-data/src/field/displayProcessor.ts

```diff
@@ -23,22 +23,39 @@ interface DisplayProcessorOptions {
   theme?: GrafanaTheme; // Will pick 'dark' if not defined
 }
 
+// Reasonable units for time
+const timeFormats: KeyValue<boolean> = {
+  dateTimeAsIso: true,
+  dateTimeAsUS: true,
+  dateTimeFromNow: true,
+};
+
 export function getDisplayProcessor(options?: DisplayProcessorOptions): DisplayProcessor {
   if (options && !_.isEmpty(options)) {
-    if (options.type && options.type === FieldType.time) {
-      return (value: any) => {
-        let dateFormat = DEFAULT_DATE_TIME_FORMAT;
-        if (options.config && options.config.dateDisplayFormat) {
-          dateFormat = options.config.dateDisplayFormat;
+    const field = options.config ? options.config : {};
+
+    if (options.type === FieldType.time) {
+      if (field.unit && timeFormats[field.unit]) {
+        // Currently selected unit is valid for time fields
+      } else {
+        const dateFormat = field.dateDisplayFormat || DEFAULT_DATE_TIME_FORMAT;
+
+        // UTC or browser based timezone
+        let fmt = (date: DateTime) => date.format(dateFormat);
+        if (options.isUtc) {
+          fmt = (date: DateTime) => toUtc(date).format(dateFormat);
         }
-        const formatedDate = getTimeZoneDateFormatter(options.isUtc ? 'utc' : 'browser')(value, dateFormat);
-        return {
-          numeric: value,
-          text: formatedDate,
+
+        return (value: any) => {
+          const date: DateTime = isDateTime(value) ? value : dateTime(value);
+          return {
+            numeric: isNaN(value) ? date.valueOf() : value,
```

Comment:
not sure if this line is necessary... `valueOf()` will work if the input was a string though